Initial client

zoidyzoidzoid opened the pull request:
Closes https://github.com/yola/teamsupport-python/issues/1

Commit:
Initial client

snitch commented on setup.py:
```diff
@@ -0,0 +1,52 @@
+setup(
+    name='teamsupport',
+    version='0.1.0',
+    description="Python library for interfacing with the TeamSupport API",
```
Q000 Remove Double quotes.

snitch commented on setup.py:
```diff
@@ -0,0 +1,52 @@
+    version='0.1.0',
+    description="Python library for interfacing with the TeamSupport API",
+    long_description=readme + '\n\n' + changelog,
+    author="Yola Engineers",
```
Q000 Remove Double quotes.

snitch commented on setup.py:
```diff
@@ -0,0 +1,52 @@
+        'Intended Audience :: Developers',
+        'License :: OSI Approved :: BSD License',
+        'Natural Language :: English',
+        "Programming Language :: Python :: 2",
```
Q000 Remove Double quotes.

snitch commented on teamsupport/utils.py:
```diff
@@ -0,0 +1,22 @@
+    if not (data or cdata):
+        raise TypeError(
+            "to_xml() needs either a 'data' or 'cdata' argument "
+            "(neither given)")
```
Q000 Remove Double quotes.

zoidyzoidzoid replied:
No.

snitch commented on tests/__init__.py:
```diff
@@ -0,0 +1,36 @@
+
+from lxml import etree
+from mock import Mock, patch
+from requests import Session, Response
```
I101 Imported names are in the wrong order. Should be Response, Session

adrianmoisey commented on requirements.txt:
```diff
@@ -0,0 +1,5 @@
+funcsigs==0.4
+lxml==3.4.4
+mock==1.3.0
+# pbr==1.5.0
```
O_o

stefanor commented on integration_tests/__init__.py:
```diff
@@ -0,0 +1 @@
+__author__ = 'william'
```
Mmmm?

stefanor commented on setup.py:
```diff
@@ -0,0 +1,52 @@
+    classifiers=[
+        'Development Status :: 2 - Pre-Alpha',
+        'Intended Audience :: Developers',
+        'License :: OSI Approved :: BSD License',
```
BSD != MIT

stefanor commented on teamsupport/__init__.py:
```diff
@@ -0,0 +1,10 @@
+__email__ = 'engineers@yola.com'
+__version__ = '0.1.0'
+
+__all__ = (TeamSupportService)
```
Missing comma?

stefanor commented on teamsupport/services.py:
```diff
@@ -0,0 +1,81 @@
+
+class TeamSupportService(XMLHTTPServiceClient):
+    def __init__(self, org_id, auth_token, url=None, **kwargs):
+        url = url or 'https://app.teamsupport.com/api/xml/'
```
Why would one want to set a URL?

zoidyzoidzoid replied:
If they had a sandbox or something it'd be useful, but while they don't, I'll drop it.

stefanor commented on teamsupport/utils.py:
```diff
@@ -0,0 +1,22 @@
+            child = etree.SubElement(xml, k)
+            child.text = v
+
+    if cdata:
```
Urgh. Whether a string is literal in XML, or in a CDATA block, should be entirely transparent to the consumer. If their API requires something to be in a CDATA block, they're idiots.

zoidyzoidzoid replied:
I took this code from the flow we had for creating XML to send to HelpSpot where we created CDATA blocks for URLs.

stefanor commented on teamsupport/utils.py:
```diff
@@ -0,0 +1,22 @@
+
+    if data:
+        for k, v in data.iteritems():
+            child = etree.SubElement(xml, k)
```
So, we never need to create documents with much nesting?

zoidyzoidzoid replied:
All the flow I've worked with now, I haven't had to.

zoidyzoidzoid replied:
I'm only ever `POST`/`PUT`ing single `Action` or `Ticket` objects.

stefanor commented:
Do you plan to add models, so that consumers don't need to deal with lxml objects?

zoidyzoidzoid commented:
Models could make life a little easier. There LXML objects have so many random fields, I was thinking of leaving out models for now, but if we just do some `getattr` magic it won't be too bad.

Commit:
License classifier should match license

Commit:
Drop unintentnional __author__

Commit:
__all__ should be an tuple

Commit:
Allow sending lxml Elements

Commit:
Drop duplicate PatchedSessionXMLTests class

Commit:
lxml wants explicit None checks

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Drop unncessary separate CDATA support

Commit:
Drop unncessary support for custom URLs

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Import order

Commit:
Add basic models for Tickets and Actions

snitch commented on teamsupport/models.py:
```diff
@@ -0,0 +1,66 @@
+        elif not (ticket_id or data):
+            raise TypeError(
+                "__init__() needs either a 'ticket_id' or 'data' argument "
+                "(neither given)")
```
Q000 Remove Double quotes.

zoidyzoidzoid replied:
No.

stefanor replied:
Sounds like a `ValueError`, rather than a `TypeError`?

zoidyzoidzoid replied:
I used `TypeError` because it's what Python raises if you give a function 1 argument when it needs 2, but `ValueError` makes more sense.

```
In [1]: class A(object):
   ...:     def __init__(self, b):
   ...:         self.b = b
   ...:

In [2]: A()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-3cd318a12eea> in <module>()
----> 1 A()
TypeError: __init__() takes exactly 2 arguments (1 given)
```

stefanor commented on CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+# Changelog
+
+## History
```
Is this block useful? Note, it's at the same level as the versions below.

stefanor commented on README.md:
```diff
@@ -0,0 +1,29 @@
+[demands](https://github.com/yola/demands).
+
+* Free software: MIT license
+* Documentation: https://teamsupport-python.readthedocs.org.
```
I don't see any sphinx documentation

stefanor commented on README.md:
```diff
@@ -0,0 +1,29 @@
+
+## Features
+
+* Client methods parse response content in order to return LXML Element objects
```
Update this, and below usage, to include models?

zoidyzoidzoid replied:
Oh yeah, this was from before I added models.

stefanor commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+demands==1.0.6
+funcsigs==0.4
```
Used?

zoidyzoidzoid replied:
by `mock`, which doesn't explicitly require it.

zoidyzoidzoid replied:
This list is also missing `property-caching` and `querylist`

stefanor commented on setup.py:
```diff
@@ -0,0 +1,52 @@
+try:
+    from setuptools import setup
+except ImportError:
+    from distutils.core import setup
```
Does this branch even work?

zoidyzoidzoid replied:
I haven't tested it. It's from a `cookiecutter-pypackage` template, so I'd expect it to work.

stefanor commented on teamsupport/models.py:
```diff
@@ -0,0 +1,66 @@
+    def __init__(self, client, ticket_id=None, data=None):
+        self.client = client
+        if data is not None:
+            self._data = data
```
If you move this out of the if block, you can get rid of the `_data = None` above.

stefanor commented on teamsupport/models.py:
```diff
@@ -0,0 +1,66 @@
+
+
+class Action(object):
+    data = None
```
Overridden by a method, below.

stefanor commented on teamsupport/models.py:
```diff
@@ -0,0 +1,66 @@
+    @property
+    def data(self):
+        if self._data is None:
+            self._data = self.client.get_ticket(self.id)
```
Are we gaining much by this indirection? We could call `get_ticket` in `__init__` and things would be a bit simpler.

zoidyzoidzoid replied:
It was more for delaying the call, but I'll move it.

stefanor commented on teamsupport/services.py:
```diff
@@ -0,0 +1,81 @@
+            if not isinstance(data, etree._Element):
+                data = to_xml(
+                    request_params['root'],
+                    request_params.get('data'))
```
We have a `data` variable.

stefanor commented on teamsupport/services.py:
```diff
@@ -0,0 +1,81 @@
+        data_set = request_params.get('data') is not None
+        if request_params.get('send_as_xml') and data_set:
+            data = request_params['data']
+            if not isinstance(data, etree._Element):
```
Maybe this should check if it's a dict, rather than checking if it's not an XML element?

I think that makes what's going on slightly clearer. At first I thought this was turning XML strings into trees, and then back into strings.

stefanor commented on teamsupport/services.py:
```diff
@@ -0,0 +1,81 @@
+                pretty_print=True)
+            request_params.setdefault('headers', {})['Content-Type'] = (
+                'application/xml'
+            )
```
That is probably shorter if broken into two lines.

Commit:
Update classifiers and documentation

Commit:
Drop unused distutils support

Commit:
Add explicit MissingArgumentError

Commit:
Tidy up _format_xml_request

Commit:
Fetch model data during initialisation

snitch commented on teamsupport/models.py:
```diff
@@ -0,0 +1,57 @@
+        elif not (ticket_id or data):
+            raise MissingArgumentError(
+                "__init__() needs either a 'ticket_id' or 'data' argument "
+                "(neither given)")
```
Q000 Remove Double quotes.

stefanor commented on teamsupport/models.py:
```diff
@@ -0,0 +1,57 @@
+
+    def __getattr__(self, name):
+        if self.data.find(name) is not None:
+            return self.data.find(name)
```
Would it make sense to `.text` here?

`self.TicketID.text` doesn't seem like a beautiful interface to me.

Commit:
getattr using .text on Elements is a lot cleaner

stormf commented on .gitignore:
```diff
@@ -0,0 +1,46 @@
+# Translations
+*.mo
+
+# Mr Developer
```
It's a bit weird seeing dev tools that aren't actually used in the project being gitignored. Maybe strip it down?

zoidyzoidzoid replied:
Yeah, i'll get rid of a bunch of things that it doesn't need to ignore for now.

adrianmoisey commented on setup.py:
```diff
@@ -0,0 +1,48 @@
+    readme = readme_file.read()
+
+with open('CHANGELOG.md') as changelog_file:
+    changelog = changelog_file.read().replace('# Changelog', '')
```
I'm assuming this isn't needed anymore?

zoidyzoidzoid replied:
Yeah, i think I'm gonna change these markdown files to rst, so it looks better on PyPI.

Commit:
Convert markdown to rst and exclude tests for final package

Commit:
Get version from init.py without import

`teamsupport`'s requirements mean it can't easily be imported in
`setup.py` without the outer virtualenv/python having the requirements
installed, and the only value that changed often was the version,
so I used the code @mitsuhiko's `click` uses for reading the
`__version__` from the `__init__.py` file.

Commit:
Trim gitignore to reduce bike-shedding

Commit:
Add models example to the README

Commit:
Add convenience imports for models

Commit:
Drop random bullet point

Commit:
Use env instead of python for testing envs

This is the way tools like msabramo's python-panci do the generation of
travis YAMLs from tox.ini.

stefanor commented on setup.py:
```diff
@@ -0,0 +1,57 @@
+
+test_requirements = [
+    'mock',
+    'property-caching',
```
What about the models? (It's not just a test requirement)

zoidyzoidzoid replied:
Fixed in https://github.com/yola/teamsupport-python/commit/a137fbf63ab84cd124dfe62375e7fd83adcfcd81

stefanor commented on teamsupport/models.py:
```diff
@@ -0,0 +1,57 @@
+        self._data = data
+        if ticket_id:
+            self._data = self.client.get_ticket(ticket_id)
+        elif not (ticket_id or data):
```
it's impossible for `ticket_id` to evaluate true, here

zoidyzoidzoid replied:
Dropped in https://github.com/yola/teamsupport-python/commit/e84b06f464e595a61d1cb1f03186589089f72275

stefanor commented on teamsupport/models.py:
```diff
@@ -0,0 +1,57 @@
+
+    @property
+    def data(self):
+        return self._data
```
Is the separate `_data` serving any purpose?

zoidyzoidzoid replied:
No, dropped in https://github.com/yola/teamsupport-python/commit/e84b06f464e595a61d1cb1f03186589089f72275

stefanor commented on tests/test_models.py:
```diff
@@ -0,0 +1,95 @@
+
+class TestTicket(BaseTeamSupportServiceCase):
+    def test_initialisation_with_id(self):
+        self.response.content = """<Ticket><TicketID>ID</TicketID></Ticket>"""
```
Triple-quote strings without newlines in them seem unnecessary

zoidyzoidzoid replied:
Fixed in https://github.com/yola/teamsupport-python/commit/b55333c8f7563d85a9af367d32007c0b2bda6487

Commit:
Clean up models

Dropped the data property function and parts of conditionals
that will never be evaluated differently

Commit:
property-caching is an instal requirement

Commit:
Drop unnecessary triple quotes

snitch commented on teamsupport/models.py:
```diff
@@ -0,0 +1,49 @@
+        elif not self.data:
+            raise MissingArgumentError(
+                "__init__() needs either a 'ticket_id' or 'data' argument "
+                "(neither given)")
```
Q000 Remove Double quotes.

Commit:
Make Ticket.actions use querylist

adrianmoisey commented on CHANGELOG.rst:
```diff
@@ -0,0 +1,7 @@
+Changelog
+========
```
Not enough `=`'s

adrianmoisey commented on setup.cfg:
```diff
@@ -0,0 +1,5 @@
+[metadata]
+description-file = README.md
```
Should this be `README.rst`?

Commit:
Whitespace

Commit:
Tidy up .rst aligning markdown

Commit:
drop unused metadata setup.cfg section

Commit:
Lint travis YAML

Commit:
Add Yola Slack notification

Commit:
Use Travis containerisation and caching

Commit:
Linting

snitch commented:
:sparkles: No lint errors found. :sparkles:

stefanor commented on CONTRIBUTING.rst:
```diff
@@ -0,0 +1,130 @@
+#. Fork the ``teamsupport-python`` repo on GitHub.
+#. Clone your fork locally:
+
+   ::
```
Style: The `::` can appear at the end of the previous paragraph.

stefanor commented on setup.py:
```diff
@@ -0,0 +1,57 @@
+]
+
+test_requirements = [
+    'mock',
```
Do we not need a funcsigs here too?

zoidyzoidzoid replied:
Yeah, we do. Though that's more because of an internal issue where we can't use mock 1.3.0, but I guess other people might encounter the same issue when using 1.2.0.

zoidyzoidzoid commented on requirements.txt:
```diff
@@ -0,0 +1,4 @@
+demands==1.0.6
```
These probably should be minimum version requirements instead of exact requirements, and we should also specify the same minimum versions in `setup.py`

stefanor commented on tests/test_services.py:
```diff
@@ -0,0 +1,72 @@
+        request_params = self.client._format_xml_request(request_params)
+        root_element = etree.Element('OuterField')
+        sub_element = etree.SubElement(root_element, 'Field1')
+        sub_element.text = 'Test field'
```
`lxml.builder` is so much prettier

stefanor commented:
:+1: in general. All I have commented on is style improvements.

I notice the models are read-only. Are the plans to allow making changes with them?

zoidyzoidzoid commented:
I wasn't planning on adding more to it now. I added support to the `send_as_xml` stuff so at least lxml Elements can be sent at the moment instead of a Python dictionary.

I can add some saving pretty simply, so I will do that in the morning.

Commit:
:: is tidier than :\n\n\t::

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Change requirements to minimum versions instead of exact

Commit:
Add Python 3 support

pypy3 doesn't work with lxml yet:
https://bugs.launchpad.net/lxml/+bug/1377501

Commit:
Fix styling test env for tox

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Use lxml builder where possible

snitch commented:
:sparkles: No lint errors found. :sparkles:

Commit:
Factor out some repeated test preparation

snitch commented:
:sparkles: No lint errors found. :sparkles:

zoidyzoidzoid commented:
I'm having issues with writing to TeamSupport from the models. I'm gonna leave it out and figure it out later.

stefanor commented on CHANGELOG.rst:
```diff
@@ -0,0 +1,7 @@
+Changelog
+=========
+
+0.1.0 (2015-09-02)
```
Don't forget to bump

stefanor commented:
:+1:

Commit:
Bump date in CHANGELOG

snitch commented:
:sparkles: No lint errors found. :sparkles:
